packages/core/src/routes/well-known/well-known.openapi.ts

```diff
@@ -0,0 +1,259 @@
+      pathMap,
+      tags,
+      ctx.request.origin
+    );
+
+    const data = supplementDocuments.reduce<OpenAPIV3.Document>(
+      (document, supplement) =>
+        deepmerge<OpenAPIV3.Document, DeepPartial<OpenAPIV3.Document>>(document, supplement, {
+          arrayMerge: mergeParameters,
+        }),
+      baseDocument
+    );
+
+    pruneSwaggerDocument(data);
+
+    if (EnvSet.values.isUnitTest) {
+      getConsoleLogFromContext(ctx).warn('Skip validating swagger document in unit test.');
+    }
+    // Don't throw for integrity check in production as it has no benefit.
+    else if (shouldThrow()) {
+      for (const document of supplementDocuments) {
+        validateSupplement(baseDocument, document);
+      }
+      validateSwaggerDocument(data);
+    }
+
+    ctx.body = {
+      ...data,
+      tags: data.tags?.slice().sort((tagA, tagB) => tagA.name.localeCompare(tagB.name)),
+    };
+
+    return next();
+  });
+
+  router.get('/.well-known/experience.openapi.json', async (ctx, next) => {
+    const experienceApiRoutes = experienceRouters.flatMap<RouteObject>((experienceRouter) => {
+      return experienceRouter.stack
+        .filter(({ path }) => !path.includes('.*'))
+        .flatMap<RouteObject>(({ path: routerPath, stack, methods }) =>
+          methods
+            .map((method) => method.toLowerCase())
+            .filter((method): method is OpenAPIV3.HttpMethods => method !== 'head')
+            .map((httpMethod) => {
+              const path = normalizePath(routerPath);
+              const operation = buildOperation(httpMethod, stack, routerPath, false);
+
+              return {
+                path,
+                method: httpMethod,
+                operation,
+              };
+            })
+        );
+    });
+
+    const pathMap = new Map<string, OpenAPIV3.PathItemObject>();
+    const tags = new Set<string>();
+
+    // Group routes by path
+    for (const { path, method, operation } of experienceApiRoutes) {
+      if (operation.tags) {
+        // Collect all tags for sorting
+        for (const tag of operation.tags) {
+          tags.add(tag);
+        }
+      }
+      pathMap.set(path, { ...pathMap.get(path), [method]: operation });
+    }
+
+    // Find supplemental documents
+    const routesDirectory = await findUp('routes', {
+      type: 'directory',
+      cwd: fileURLToPath(import.meta.url),
+    });
+    assertThat(routesDirectory, new Error('Cannot find routes directory.'));
+
+    const supplementPaths = await findSupplementFiles(routesDirectory, {
+      includeDirectories: ['experience', 'interaction'],
+    });
+
+    const allSupplementDocuments = await Promise.all(
+      supplementPaths.map(async (path) =>
+        removeUnnecessaryOperations(
+          // eslint-disable-next-line no-restricted-syntax -- trust the type here as we'll validate it later
+          JSON.parse(await fs.readFile(path, 'utf8')) as DeepPartial<OpenAPIV3.Document>
+        )
+      )
+    );
```

Comment:
copy - paste here? can we do a util?

Reply:
Updated extract a `getSupplementDocuments` util method